### DISCUSSION TITLE
Sinatra 3 release

### DISCUSSION THREAD

**Discussion by jkowens** [sinatra/sinatra#1789] [General]
Hi @olleolleolle, @dentarg, @epergo, @raxoft! I now have access to push up to RubyGems. I would like to finally cut a 3.0 release, but is there anything else we should get merged into main first?

**Comment by olleolleolle** [comment]
https://github.com/sinatra/sinatra/pull/1677 I like this one for 3.0 - I can ask around for PR takers!

**Reply by olleolleolle** [reply]
I got hold of @ioquatix who will look into it.

**Reply by jkowens** [reply]
That's great thanks Olle 👍

**Reply by olleolleolle** [reply]
https://github.com/sinatra/sinatra/pull/1794 @jkowens 

Can you please Approve the CI build on this thing?

**Reply by jkowens** [reply]
It turns out that we can add Falcon support, but users will see an error message when shutting the server down in development. The fix will have to be in Falcon. Should we continue merging this now?

**Reply by olleolleolle** [reply]
The stopping, @ioquatix had some ideas about that, adding a wrapper around stop, that each adapter could implement, and Falcon implement, too, in a way which dodges the issue you mentioned.

**Reply by jkowens** [reply]
Yeah having a wrapper would be a big improvement from the current implementation.

**Reply by ioquatix** [reply]
We will try to address this root interface issue in Rack itself. The `Rack::Handler` does not have a well defined interface. The `server` that gets yielded from `Rack::Handler::(ServerClass).run` isn't well defined and there wasn't an expectation that it has as `#stop` method or any consistent interface at all. It makes sense to me that this is better defined, but I'm not sure this ultimately belongs to "Rack the interface" as this is more "Rack the server implementations". https://github.com/rack/rack/discussions/1928

The simplest solution is for Falcon just to support the defacto "#stop" method and we can provide a wrapper from `Rack::Handler::Falcon.run` which supports this interface in the Falcon gem - I'll work on that today, and it won't impact the 3.0 release of Sinatra - you can merge the PR and I'll fix it on my end. There isn't any extra work that needs to be done in sinatra right now IMHO.

Going forward, you might like to evaluate HOW sinatra start servers. The `Rack::Handler` infrastructure isn't well defined but I think we should improve it. Looking to sinatra, I see two parts: starting the server, and handling interrupts to stop it. But this makes a number of assumptions about when it's safe to call #stop. Calling `#stop` from signal handlers might not be safe - I'll make sure Falcon supports that, but I'm not sure about other servers. Calling webrick `#stop` from a signal handler probably isn't safe - it might be, but I wouldn't bet on it!

**Reply by ioquatix** [reply]
Falcon v0.40.1 is released and should work well with Sinatra.

**Reply by jkowens** [reply]
Wow thank you @ioquatix!!

**Comment by epergo** [comment]
That's great @jkowens ! It would be nice if we could finish https://github.com/sinatra/sinatra/issues/1759 first because they are breaking changes, 3.0 is a great opportunity to include those.

**Reply by jkowens** [reply]
@epergo good point, that one is close!

**Reply by raxoft** [reply]
It's a shame SASS had to go. We use SASS, Coffeescript and Slim in our apps. But fortunately the first 2 are used only for convenience for the on-the-fly compilation during development, for deployment we use precompiled resources anyway, so it shouldn't harm us that much.

**Reply by dentarg** [reply]
SASS could always be added back, if someone takes responsibility for it and contributes. Are you using any of the the not so maintained gems ([ruby-sass](https://github.com/sass/ruby-sass), [sassc-ruby](https://github.com/sass/sassc-ruby)) @raxoft or something else? Just saw https://github.com/ntkme/sassc-embedded-shim-ruby + https://github.com/ntkme/sass-embedded-host-ruby (linked in https://github.com/sass/sassc-ruby/issues/220)

**Reply by raxoft** [reply]
We are using `sassc-ruby` ATM. But it's true that that game looks abandoned for quite some time now, so perhaps switching to something else might help. However I am leaving for two weeks vacation in couple of days, so I am afraid I can't explore those options now. But as you said, it can always be added later again. It's just that my line of thinking was that when building an app, in general you need HTML templates, CSS and JS. So it would be nice if sinatra ships with something for each of these out of the box. :man_shrugging:

**Reply by dentarg** [reply]
I think people are starting to question if you still need SASS, example: https://github.com/thoughtbot/administrate/pull/2116#issuecomment-1036110366

also, sure, that would be nice but the reality is that Sinatra is barley alive as a project and few have incitement to develop and maintain it for others, for free

**Reply by epergo** [reply]
> It's just that my line of thinking was that when building an app, in general you need HTML templates, CSS and JS. So it would be nice if sinatra ships with something for each of these out of the box

You have a point here, however I think that instead of supporting sass, coffescript, and N engines for assets a gem that is specialized in working with assets and precompiling them for production like [Sinatra Asset Pipeline](https://github.com/kalasjocke/sinatra-asset-pipeline) (note: I haven't used it) could be used instead

**Comment by jkowens** [comment]
It seems like we might be ready to cut the 3.0 release. Anything else to consider? 😅

**Reply by ioquatix** [reply]
Rack 3 will be released soon and will change a number of things. A 3.0 release of Sinatra is great, and it will live for a long time I am sure, be aware that enough will be changing that I imagine you will want to make a Sinatra 4.0 for Rack 3+. Just a heads up and maybe worth considering for the longer term plan.

**Comment by jkowens** [comment]
Maybe we should see this PR through: https://github.com/sinatra/sinatra/pull/1692

I'm trying to decipher all the history/comments 😄  If anyone has input on that one, it would be appreciated!

**Reply by dentarg** [reply]
I stopped using `sinatra/reloader` years ago, as it wasn't reliable(or it doesn't reload everything you might change in a project (models) so you need to restart anyway?). I wouldn't spend any time on it, I think it should be deprecated. The Sinatra project doesn't seem to have the bandwidth to maintain such feature :/

**Reply by jkowens** [reply]
@dentarg this prompted me to look at current deprecation warnings. It looks like there are several, this is probably a good time to delete that code.

I think your suggestion about reloader is good. I think we should deprecate with v3 and then plan on removing in v4.

**Reply by ioquatix** [reply]
Reloading, IIUC, should be the job of the server, not the application middleware.

**Reply by jkowens** [reply]
@dentarg any suggestions on how to go about deprecating these sinatra-contrib extensions? I'm not sure what to do exactly. Should we just document them as deprecated? Then would they eventually just be deleted, or do they get pulled into another repo to be archived?

**Reply by dentarg** [reply]
@jkowens Like how it was done for "decompile"? It can still be loaded but only thing it does is print a warning. https://github.com/sinatra/sinatra/commit/e48eac3291598e751e71e89f61ef0fda811a5127#diff-dbd6906666d08a2c144fe245966fdff82eff3bc9d94982b2448898b44ffe5184

If there's demand for it, it could always be extracted to a separate repo later, to make it easier for people still wanting to use it.

**Reply by jkowens** [reply]
I did see that for decompile, but that looks more like immediate removal with a warning than deprecation 😅

**Reply by epergo** [reply]
We can extract deprecated extensions to their own repository now, add them as dependencies of sinatra-contrib (so everything keeps working as it is now) and show a warning when requiring them.

Then in later releases of contrib we can drop that dependency.

**Reply by dentarg** [reply]
@jkowens True but if we're talking Sinatra 3 my thinking was we can make such breaking changes :)

**Reply by jkowens** [reply]
Very true, I just think we should generally allow people to upgrade and still use deprecated extensions if they choose. Maybe that isn't necessary for reloader, but I think we might want to do that if any of the other extensions need to be deprecated.

I'm leaning towards updating the rdoc (and other documentation) to make it clear reloader is deprecated. Then in the v4 release just delete it 😅

**Comment by jkowens** [comment]
Hey so I decided to try to tackle the Rubocop PR #1537 and see if that could be merged. I got it rebased and updated, but I'm seeing a TruffleRuby failure. Hopefully it's not too hard to track down.

**Reply by jkowens** [reply]
Ah I found the cop that was the issue. If anyone is able to give that PR another look that would be appreciated 👍

**Comment by jkowens** [comment]
I'm ready to cut a 3.0 release, but I can't figure out why the JRuby 9.3 test runs are failing now 🤔  Anyone have thoughts on what may be happening?

**Reply by dentarg** [reply]
Maybe due to https://github.com/ruby/setup-ruby/releases/tag/v1.113.0

Try running CI with latest bundler always

**Reply by olleolleolle** [reply]
https://github.com/sparklemotion/nokogiri/pull/2618/files Example PR which fixed it for another project.

**Reply by jkowens** [reply]
Nice, thanks @olleolleolle. It's weird, now it also looks like Tilt is causing test failures 😢

**Reply by olleolleolle** [reply]
Those look like keyword arguments arity issues.

**Reply by epergo** [reply]
It was an issue with a new release of Hamlit, details in https://github.com/sinatra/sinatra/pull/1809

**Comment by ioquatix** [comment]
Rack 3.0.0 beta1 was released.

It is likely this will be very close to the final release.

I encourage you to start testing with it.

https://rubygems.org/gems/rack/versions/3.0.0.beta1

It would be awesome if Sinatra 3.0 was compatible with Rack 3.0

We will note this compatibility on our README if confirmed.

**Reply by dentarg** [reply]
We have a "rack master" job in CI (allowed to fail). I see now that it has problems https://github.com/sinatra/sinatra/runs/7739048508?check_suite_focus=true#step:3:28

I can take a look next week if no one beats me to it

**Reply by ioquatix** [reply]
https://github.com/sinatra/sinatra/pull/1811

**Reply by jkowens** [reply]
I think it might be hard/impossible? to release a Sinatra 3 that supports both Rack 2 and 3 (due to rack-session). I'm thinking we should release Sinatra 3 with Rack 2 support and the follow up with a Sinatra 4 release that moves to Rack 3. Please let me know I'm wrong or there are other factors to consider 😄

**Reply by ioquatix** [reply]
I have my own web framework and will work through this independently and report back.

I don't know for sure, but I believe it should be possible to make `rack-session` compatible with rack 2.

I'll take a look at whether this is possible, because it would be a nice compromise/middleground.

**Reply by raxoft** [reply]
> I'm thinking we should release Sinatra 3 with Rack 2 support and the follow up with a Sinatra 4 release that moves to Rack 3. Please let me know I'm wrong or there are other factors to consider smile

Hmm, personally, if Sinatra 3.0 brings breaking changes, I would rather prefer if it is based on Rack 3.0 already, so I can upgrade both at once. Otherwise I would be inclined to wait for 4.0 anyway so I don't have to do the work twice.

**Reply by dentarg** [reply]
> Hmm, personally, if Sinatra 3.0 brings breaking changes, I would rather prefer if it is based on Rack 3.0 already, so I can upgrade both at once. Otherwise I would be inclined to wait for 4.0 anyway so I don't have to do the work twice.

I would not prefer that, I like smaller updates, also it sounds from @jkowens it isn't even possible.

I agree with this: "I'm thinking we should release Sinatra 3 with Rack 2 support and the follow up with a Sinatra 4 release that moves to Rack 3"

**Reply by jkowens** [reply]
I suspect upgrading to Sinatra 3 will be fairly trivial for most applications, so I'm leaning towards releasing Sinatra 3 as is unless there is a way to support both Rack 2 and 3. @ioquatix have you had a chance to investigate supporting both with `rack-session`?

**Reply by ioquatix** [reply]
Sinatra should probably declare it's dependency on rack >= 2 and then users should expect to add specifically what version they want in their gemfile, including a dependency on `rack-session` if needed.

If Sinatra has a hard dependency on rack-session, maybe you can release `sinatra-session` or `Sinatra-rack` which has one release depending on `rack ~> 2.0` and another release depending on `rack ~> 3.0` + `rack-session`. I'm not sure if this will work but it's my best idea.

**Reply by jkowens** [reply]
Thanks @ioquatix, I appreciate the feedback. Sinatra does have a hard dependency on rack-session. I'm getting the sense that it will just be easier to release Sinatra 3 on Rack 2 and then then move forward with Sinatra 4 on Rack 3. I think I'll aim to cut a release this weekend so we can get moving on Rack 3 support.

**Reply by ioquatix** [reply]
Whatever way you choose, thanks for your hard work.

It might be nice to get some input from Bundler/RubyGems team on whether we can put version dependent dependencies. We could also have a "null" release of `rack-session` which is pinned to Rack 2, and then have another release with the actual code with a requirement on rack 3+. If you want to explore supporting both Rack 2 and Rack 3 in Sinatra 3, I'll be happy to help figure out the options.

**Comment by jkowens** [comment]
Sinatra 3 has been quietly released 😅 

I didn't want to continue sitting on it and now we can focus on Rack 3. It seems we could possibly add support in a minor release, but if it requires a major release we can do that.

**Reply by ioquatix** [reply]
Nice! I'm happy to help. There are still some open questions about the best way to handle `rack-session`. Maybe it's best just to move forward and don't support Rack 2. Most applications should prefer Rack 3 going forward. In a world where Rack 2 and Rack 3 exist, there is not much point to support Rack 2 for frameworks.

You could consider keeping 3.0.x as Rack 2 and maintain it (bug fixes etc), but use 3.1+ as Rack 3 + new features/development.
